Add i18n support to app/myelectric in emoncms 9.x

euqip opened the pull request:
The support is only possible when enabling user language instead of 'en' with apikey login! This is to be done in user_model.php and is not part of this pull request.

euqip commented:
I think it's now complete.
PS: when using APIKEY to display data is like login in and it gives access to all kind of data the user has access to.  Is it not possible to make some restrictions?

glynhudson commented:
Thanks for this, just tested quickly I get error when I try and load emoncms after switching to your branch:

`Fatal error: Call to a member function fetch_array() on boolean in /var/www/emoncms/Modules/user/user_model.php on line 443`

https://github.com/emoncms/emoncms/blob/master/Modules/user/user_model.php#L443

euqip commented:
Hi glynhudson,

I think I had the same error before I made the database update.  I have written the README just after that episode.
App module uses a specific table to store its settings.

Beside the need to update the database, there maybe a better way to advise the user of a missing table.
I will look for it in a near future.

I see two methods:
1 send a message to advise the usser a database update need.
2 force the database update if table does not exist

Which is the best?

glynhudson commented:
I would say the database would need force update since after updating to these changes I could not even load the emoncms login page, therefore is no way to update the database in the usual way. What do you think @TrystanLea ?

euqip commented:
@glynhudson
Please before testing this code, correct the user_model.php file (around line 442) where the sql query selecting user  language is incorrect.  This error correction is not part of app module.

I added a table existence test and an error page with a direct access to 'admin/db' for a db update using built-in procedure.

TrystanLea commented:
Hello Equip

I've made the correction to user_model. 

I think your check table function needs to be:

    public function checktable(){
        // redirect to /admin/db when table does not exist
        $sql = 'SHOW TABLES LIKE "app_config"';
        $result = $this->mysqli->query($sql);
        if ($result->num_rows>0) return true;
        return false;
    }

Would you be able to set the colours for the myelectric totals all to blue? as the other colours are more in relation to solarpv etc.

In myelectric config I see the error: "Missing keys in i18n file" should that be shown at this point? could it be hidden if not needed?

TrystanLea commented:
The graph part of my heatpump view has also lost the way the background colour used to stretch to the screen width and are the value colours for solarpv and wind in my energy changed?

TrystanLea commented:
When you adjust the myelectric page width its also doing some strange things.. the realtime graph expands down the page. Is there a way to maintain the previous behaviour? I havent had a chance yet to look through the code in detail

euqip commented:
Hi Tristan and Glyn

In new pullrequest I changed the checktable code the way proposed, its shorter and efficient.
I changed the myelectric colours to blue
I added the missing canvas css rules to keep graph height and width adapted to screen width, without changing height.
In Myheatpump, I checked the original version where the canvas width is fixed to 1200px, and I have kept this width.
You recently added a README and a picture file, mine is maybe toot much.
I removed the missing keys visibility, this was the way I used to collect the texts to translate and may be reused later if any changes occur.
I normally did not change any logic.
I found that in myenergy, when the required feeds does not exist and you save the configuration, the interface hangs.  This error is not yet corrected.
When linting the app.js code, I added a bug. it is now restored (the config save did not occur)  The original JS code included global variables, undefined variables and vars defined within loops.

Hope it will be better now.

glynhudson commented:
Just tested and I'm afraid it still seems to be buggy, see attached for before and after testing your changes screengrabs:

## Before:
![screenshot 2016-02-13 at 11 54 35 - display 2](https://cloud.githubusercontent.com/assets/758844/13027463/39cb7f3c-d249-11e5-8e50-8310226f6bf1.png)

## After: 
![screenshot 2016-02-13 at 11 52 10 - display 2](https://cloud.githubusercontent.com/assets/758844/13027464/3fb595cc-d249-11e5-9f22-4c4e688a4bf1.png)

Daily values are incorrectly calculated 

## Before: 
![screenshot 2016-02-13 at 11 54 45 - display 2](https://cloud.githubusercontent.com/assets/758844/13027466/4cb19712-d249-11e5-9422-5c8c820e7db2.png)

## After: 
![screenshot 2016-02-13 at 11 52 44 - display 2](https://cloud.githubusercontent.com/assets/758844/13027465/44e56b30-d249-11e5-8bff-16ed8d18c018.png)




Formatting errors

euqip commented:
Hi Glyn,

I h

TrystanLea commented:
The solar pv page gives the following error in my browser:

TypeError: Math.parseInt is not a function

I think it should be just parseInt (https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/parseInt)

Otherwise the myelectric page is now responding correctly to screen resizes :) Although not sure if currency is working properly? it prints pound to screen rather than £, the original works as expected for me.

TrystanLea commented:
+ sorry I see that the heatpump page formatting is actually incorrect in the original too. Im working on a new interface for this so might be best for me to remove this page completely

euqip commented:
Hi Tristan
Because I introduced € (utf-8), the way the page was filled does not work the same way.

this code has been changed
```javascript
            $("#myelectric_usetoday_units_a").html("&"+app_myelectric.currency+";");
to
            $("#myelectric_usetoday_units_a").html(app_myelectric.currency);
```
Could you please enter the pound symbol (£) in your settings, and see what happens.
Before, you had &pound; displayed as a £

Hope it will work better.

I just discovered the &euro; works too. (why not), but if you choose for USD, it will not be ok.
An other way to solve it is to set &pound; or &euro; in settings to keep all freedom.

euqip commented:
@Tristan,

parseInt is gives an error in jslint, but works perfect.  After having rechecked at W3S and Mozilla, its practically the only math function not using this prefix.  I changed my source.
